### PR TITLE
docs(dev-server): clarify root path for generated files

### PR DESCRIPTION
**Pull request by maxloh**
Clarify root path for generated files

Related: https://github.com/webpack/webpack-dev-server/issues/2982

### CODE REVIEW

File: src/content/configuration/dev-server.md
Comment (alexander-akait) [inline src/content/configuration/dev-server.md:71]: I think we should clarify `publicPath` here otherwise it some misleading

Comment (maxloh) [reply src/content/configuration/dev-server.md:71]: Maybe we should mention both solutions? If the publicPath solution is adopted, the logic will be required to be defined duplicately, in both `output.path` and `devServer.publicPath`.

Comment (alexander-akait) [reply src/content/configuration/dev-server.md:71]: `webpack-dev-server@4` respect `output.publicPath`, but I think we should clarify:
- output.path + output.filename (i.e. `output.publicPath` is empty for webpack@4 and `auto` for webpack@5)
- output.path + output.publicPath + output.filename

Comment (maxloh) [reply src/content/configuration/dev-server.md]: I have rewritten the paragraph. Is it correct and OK?

**Comment by chenxsan** [comment]
Thanks for your contribution here, however I don't think it's a good idea to add such a large block of texts to the configuration page, it's better to keep it lean. We can add it to a guide like https://webpack.js.org/guides/development/, but let's keep it simple as well, the current version seems unnecessarily complex.

**Comment by maxloh** [comment]
> Thanks for your contribution here, however I don't think it's a good idea to add such a large block of texts to the configuration page, it's better to keep it lean. We can add it to a guide like https://webpack.js.org/guides/development/, but let's keep it simple as well, the current version seems unnecessarily complex.

Ok.

**Comment by maxloh** [comment]
@chenxsan @alexander-akait Updated according to your comments.

File: src/content/guides/development.md
Comment (maxloh) [reply src/content/guides/development.md:296]: How about the alternative way to setup dev-server, with `output.path` and `output.filename`?
``` diff
 module.exports = {
   output: {
     filename: 'bundle.js',
     path: path.resolve('assets'), // Keep this line intact to specify output directory for building
+    publicPath: '/assets/', // Tell webpack-dev-server to serve generated files under assets directory
   },
 };
```

File: src/content/configuration/dev-server.md
Comment (maxloh) [reply src/content/configuration/dev-server.md:31]: A tip in the API documentation page will be more easy to spot as people tends to find solutions in API documentations instead of guides.

I can remove it though if you really want.

Comment (chenxsan) [reply src/content/configuration/dev-server.md:31]: There're already a lot of tips/warnings,

image

It's so distracting in my opinion. We can't cover every basic concepts on API documentation, guides serve that purpose well.

File: src/content/guides/development.md
Comment (chenxsan) [reply src/content/guides/development.md:296]: Sorry, I'm a little confused by your example, why would you want to set the `output.publicPath` when developing? As far as I know, you can just leave it unset and everything should work just fine.

Comment (chenxsan) [reply src/content/guides/development.md:296]: Ok, I just took some time to check your example. In your original example here https://github.com/webpack/webpack-dev-server/issues/2982, no `publicPath` is provided, and it defaults to `auto`, webpack dev server would transform it into `/` as per https://github.com/webpack/webpack-dev-server/blob/788f2ba7d9152610ba0d3024f0079c98aad86991/lib/Server.js#L632, hence the `app.bundle.js` is served from `/app.bundle.js` instead of `./assets/js/app.bundle.js`. I think you should consider improve your project structure first. What we're documenting here, I would say it's bad practice, we can easily shoot ourself in the foot.

Comment (chenxsan) [reply src/content/guides/development.md:296]: Anyway, if you insist these should be added, I'll leave it to other reviewers.

File: src/content/configuration/dev-server.md
Comment (maxloh) [reply src/content/configuration/dev-server.md:31]: OK

File: src/content/guides/development.md
Comment (maxloh) [reply src/content/guides/development.md:296]: Sorry I referenced the wrong code snippet. I have just updated it.

It doesn't feel so right as defining both `output.path` and `devServer.publicPath` seems like adding duplicated logic to the config.

It is your call though as maintainers have more knowledge to their projects.

Comment (chenxsan) [reply src/content/guides/development.md]: Generally, we don't have to worry about `output.publicPath`.

Here's what I know about `output.path` and `output.publicPath`. With tools like webpack, we actually have more than one environments, for example, where we build the project could be different from where we deploy the project.

Let's say I'm going to build a project `/users/sam/project` on my computer.

1. `output.path` decides where we build the project to, e.g., `/users/sam/project/dist`
2. now I want to deploy all files under `dist` to a production server under domain `https://example.com`, I can just copy all files under local `dist` to the root of my domain, e.g., `/users/ubuntu/site_a/`. Everything's fine so far.
3. however, I change my mind later and decide to deploy my project under `https://example.com/blog/`, with the default `output.publicPath`, I can copy files to `/users/ubuntu/site_a/blog/` directory, and everything should just work fine too.
4. Some times later, I decide to host assets like image/css/js etc. on a CDN service, we copy those assets to cdn and we can access them from address like `https://cdn/what/ever/path/main.js`. However the `index.html` is still on our own server, i.e., we want the `index.html` to load assets from cdn instead of our own server, now we need to set `output.publicPath` to `https://cdn/what/ever/path/`, and the html file webpack generated will reference `https://cdn/what/ever/path/main.js` instead of `https://example.com/blog/main.js`.

Comment (maxloh) [reply src/content/guides/development.md]: OIC. webpack is designed for building all front end files at once. In most cases, the html files that referencing css and js files will also process by webpack and located at `output.path`.

In that case there won't be any duplicated logic in the config.

Comment (maxloh) [reply src/content/guides/development.md]: By the way, can you help merge the pull request if it is OK?

File: src/content/guides/development.md
Comment (chenxsan) [inline src/content/guides/development.md:269]:
```suggestion
T> `webpack-dev-server` serves bundled files from the directory defined in [`output.path`](/configuration/output/#outputpath), i.e., files will be available under `http://[devServer.host]:[devServer.port]/[output.publicPath][output.filename]`.
```

**Review by chenxsan — APPROVED** [review]

**Pull request merged**
